en/2-wallet/README.md

```diff
@@ -14,3 +14,360 @@
+
+This is how secp256k1 curve looks:  
+![](https://cdn-images-1.medium.com/max/1600/1*4dcCrlQfGqZECDLy-25fnw.png)
+
+First, we'll explain how to do point addition on an elliptic curve.  
+Having points *P* and *Q*, we draw a line that goes through both of them and find third point on the intersection of the curve and the line. Then we reflect that point across x-axis to get *R*.  
+*P + Q = R*  
+![](https://cdn-images-1.medium.com/max/1600/1*dErGh_rL2ExM6AX-Rtyb7w.png)
+
+But in the elliptic cryptography, rather than adding two arbitrary points together, we add specified base point on the curve to itself.  
+We draw tangent line to the curve at the point *P*, then we apply the same rules as above.  
+*P + P = 2 * P*  
+![](https://cdn-images-1.medium.com/max/1600/1*ffYKgW-4_Paxve3G1HIJXw.png)
+
+Remember that private key is a 256-bit number? Basically, public key is the result of *P* adding to itself *x* times, where *x* is our private key.  
+
+*X = x * P*, where *X* is the public key.
+
+To do the opposite (figure out *x* from *X* and *P*) we have to keep adding *P* to itself until we get *X*, which on average make us do *2^128* point additions to figure out *x*, which is computationally infeasible.
+
+## ECDSA signing
+### Signature calculation
+To calculate signature for arbitrary data we do the following:  
+1. Calculate *z = sha256(data)* which is interpreted as 256-bit integer
+2. Choose random integer *k* between *1* and *n - 1*, where *n* is *115792089210356248762697446949407573529996955224135760342422259061068512044369* for secp256r1
+3. Calculate point *P(x, y) = k * G*, where *G* is base point (48439561293906451759052585252797914202762949526041747995844080717082404635286,
+36134250956749795798585127919587881956611106672985015071877198253568414405109) for secp256r1
+4. Calculate *r = x mod n*. If *r = 0*, then try again from the start.
+5. Calculate *s = k^−1 * (z + r * d) mod n*, where *d* is our private key. If *s = 0*, then try again from the start.
+
+*k^-1* is actually a modular multiplicative inverse of *k*, which is a number such that *(k^-1 * k) mod n = 1*
+
+Pair *(r, s)* is the signature. Usually they're concatenated together to a single hexstring.  
+
+### Signature verification
+Signature verification is based on the ability to calculate point *P* from above w/o actually knowing *k* or *d*.
+
+1. Calculate *u = s^−1 * z mod n*
+2. Calculate *v = s^−1 * r mod n*
+3. Calculate point *P(x, y) = u * G + v * Q*, where *Q* is our public key
+4. Verify that *r = x mod n*, otherwise signature is invalid
+
+### Security
+It's very important to use cryptographycally secure random *k* every time, otherwise having two signatures *s* and *s'*, generated with the same *k* we can easily find out *k*:  
+*k = (z - z') / (s - s') mod n*, where *z* and *z'* are hashes of signed data  
+
+If we know *k* then we can easily find out private key:  
+*d = (s * k - z) / r mod n*
+
+## NEO Address
```

Comment:
I think the expression here is inappropriate.
NEO address is generated from a script(We often call it address script).And its fomat is  
0x21 (1 byte, representing OptCode.PUSHBYTES21) + compressed public key (33 bytes) + 0xac (1 byte, representing OptCode.CHECKSIG)